src/main.cc

```diff
@@ -296,8 +318,653 @@ uint64_t rotr64 (uint64_t n, unsigned int c)
     return (n>>c) | (n<<( (-c)&mask ));
 }
 
+/*
+Georgios's Page Walker
+Author: Georgios Vavouliotis
+Advisors: Lluc Alvarez, Marc Casas Guix
+{georgios.vavouliotis, lluc.alvarez, marc.casas}@bsc.es
+
+Ported by James Coman - james in the domain tamu.edu
+*/
+int simulated_page_walker(uint32_t cpu, uint64_t vpage, int swap, uint8_t is_code, uint64_t instr_id, uint64_t ip, int type){
+
+	// 4KB: LOG2_PAGE_SIZE == 12
+	// 2MB: LOG2_PAGE_SIZE == 21
+
+	int cstall = 2; // the latency cost of the current page walk
+
+	// we let ChampSim decide if an access is a swap or not... If it is not the simulated page walk hardware is activated
+	// we also use the translation that ChampSim generates but we follow a realistic page walk procedure to get the latency cost of each page walk
+    if (swap == 0){
+
+    	ooo_cpu[cpu].STLB.mmu_timer++;
+
+    	// searching the MMU cache levels for possible hits
+    	int mmu_hit[3] = {0,0,0};
+    	uint64_t compare_pml4, compare_pdp, compare_pd;
+
+    	if(LOG2_PAGE_SIZE == 12){
+        	compare_pml4 = vpage>>27;
+        	compare_pdp  = vpage>>18;
+        	compare_pd   = vpage>>9;
+    	}
+    	else{
+        	compare_pml4 = vpage>>18;
+        	compare_pdp  = vpage>>9; 
+    	}
+
+    	mmu_hit[0] = ooo_cpu[cpu].STLB.search_pml4(compare_pml4); // check for hits in the MMU cache of the PML4 level
+    	mmu_hit[1] = ooo_cpu[cpu].STLB.search_pdp(compare_pdp);   // check for hits in the MMU cache of the PDP level
+
+    	// if you use 2MB pages there is one page table level less to search on
+    	if(LOG2_PAGE_SIZE == 12)
+        	mmu_hit[2] = ooo_cpu[cpu].STLB.search_pd(compare_pd); // check for hits in the MMU cache of the PD level
+
+    	// update the MMU cache levels
+        ooo_cpu[cpu].STLB.update_pml4(ooo_cpu[cpu].STLB.mmu_timer, compare_pml4); // update the content of the MMU cache of the PML4 level if needed
+        ooo_cpu[cpu].STLB.update_pdp(ooo_cpu[cpu].STLB.mmu_timer, compare_pdp);   // update the content of the MMU cache of the PDP level if needed
+        if(LOG2_PAGE_SIZE == 12)
+        	ooo_cpu[cpu].STLB.update_pd(ooo_cpu[cpu].STLB.mmu_timer, compare_pd); // update the content of the MMU cache of the PD level if needed
+
+		uint64_t cr3 = 0x200000; // defining the CR3 register pointing address 
+        uint64_t pt_index, pd_index, pdp_index, pml4_index;
+
+		// grouping the v@ bits ... note that for 2MB pages there is one page table level lookup less
+        if(LOG2_PAGE_SIZE == 12){
+            pt_index   = (vpage       & 0x00000000001ff);
+            pd_index   = ((vpage>>9)  & 0x00000000001ff);
+            pdp_index  = ((vpage>>18) & 0x00000000001ff);
+            pml4_index = ((vpage>>27) & 0x00000000001ff);
+        }
+        else{
+            pt_index   = (vpage       & 0x00000000001ff);
+            pdp_index  = ((vpage>>9)  & 0x00000000001ff);
+            pml4_index = ((vpage>>18) & 0x00000000001ff);
+        }
```

Comment:
The same flag is used multiple times here. So I would suggest the same as in the previous comment. Set it in a `#define` and that would look much cleaner. 😉